flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/testutils/ValidatingExactlyOnceSink.java

```diff
@@ -33,23 +34,29 @@
 
 /** A {@link RichSinkFunction} that verifies that no duplicate records are generated. */
 public class ValidatingExactlyOnceSink extends RichSinkFunction<Integer>
-        implements ListCheckpointed<Tuple2<Integer, BitSet>>, Runnable {
+        implements ListCheckpointed<Tuple2<Integer, BitSet>>, Runnable, CheckpointListener {
 
     private static final Logger LOG = LoggerFactory.getLogger(ValidatingExactlyOnceSink.class);
 
     private static final long serialVersionUID = 1748426382527469932L;
 
     private final int numElementsTotal;
+    private final boolean waitForFinalCheckpoint;
 
     private BitSet duplicateChecker = new BitSet(); // this is checkpointed
 
-    private int numElements; // this is checkpointed
+    private volatile int numElements; // this is checkpointed
```

Comment:
I think you do not need to make this volatile. Flink guarantees single-threaded-style invocations of all methods, so the visibility in different methods is guaranteed.